src/Data/Vec.agda

```diff
@@ -8,270 +8,42 @@
-
--- Special folds
-
-sum : ∀ {n} → Vec ℕ n → ℕ
-sum = foldr _ _+_ 0
-
-count : ∀ {P : Pred A p} → Decidable P → ∀ {n} → Vec A n → ℕ
-count P? []       = zero
-count P? (x ∷ xs) with P? x
-... | yes _ = suc (count P? xs)
-... | no  _ = count P? xs
-
-------------------------------------------------------------------------
--- Operations for building vectors
-
-[_] : A → Vec A 1
-[ x ] = x ∷ []
-
-replicate : ∀ {n} → A → Vec A n
-replicate {n = zero}  x = []
-replicate {n = suc n} x = x ∷ replicate x
-
-tabulate : ∀ {n} → (Fin n → A) → Vec A n
-tabulate {n = zero}  f = []
-tabulate {n = suc n} f = f zero ∷ tabulate (f ∘ suc)
-
-allFin : ∀ n → Vec (Fin n) n
-allFin _ = tabulate id
-
-------------------------------------------------------------------------
--- Operations for dividing vectors
-
-splitAt : ∀ m {n} (xs : Vec A (m + n)) →
-          ∃₂ λ (ys : Vec A m) (zs : Vec A n) → xs ≡ ys ++ zs
-splitAt zero    xs                = ([] , xs , refl)
-splitAt (suc m) (x ∷ xs)          with splitAt m xs
-splitAt (suc m) (x ∷ .(ys ++ zs)) | (ys , zs , refl) =
-  ((x ∷ ys) , zs , refl)
-
-take : ∀ m {n} → Vec A (m + n) → Vec A m
-take m xs          with splitAt m xs
-take m .(ys ++ zs) | (ys , zs , refl) = ys
-
-drop : ∀ m {n} → Vec A (m + n) → Vec A n
-drop m xs          with splitAt m xs
-drop m .(ys ++ zs) | (ys , zs , refl) = zs
-
-group : ∀ n k (xs : Vec A (n * k)) →
-        ∃ λ (xss : Vec (Vec A k) n) → xs ≡ concat xss
-group zero    k []                  = ([] , refl)
-group (suc n) k xs                  with splitAt k xs
-group (suc n) k .(ys ++ zs)         | (ys , zs , refl) with group n k zs
-group (suc n) k .(ys ++ concat zss) | (ys , ._ , refl) | (zss , refl) =
-  ((ys ∷ zss) , refl)
-
-split : ∀ {n} → Vec A n → Vec A ⌈ n /2⌉ × Vec A ⌊ n /2⌋
-split []           = ([]     , [])
-split (x ∷ [])     = (x ∷ [] , [])
-split (x ∷ y ∷ xs) = Prod.map (x ∷_) (y ∷_) (split xs)
-
-------------------------------------------------------------------------
--- Operations for converting between lists
-
-toList : ∀ {n} → Vec A n → List A
-toList []       = List.[]
-toList (x ∷ xs) = List._∷_ x (toList xs)
-
-fromList : (xs : List A) → Vec A (List.length xs)
-fromList List.[]         = []
-fromList (List._∷_ x xs) = x ∷ fromList xs
-
-------------------------------------------------------------------------
--- Operations for reversing vectors
-
-reverse : ∀ {n} → Vec A n → Vec A n
-reverse {A = A} = foldl (Vec A) (λ rev x → x ∷ rev) []
-
-infixl 5 _∷ʳ_
+-- Additional operations
 
-_∷ʳ_ : ∀ {n} → Vec A n → A → Vec A (1 + n)
-[]       ∷ʳ y = [ y ]
-(x ∷ xs) ∷ʳ y = x ∷ (xs ∷ʳ y)
+module _ {P : A → Set p} (P? : Decidable P) where
 
-initLast : ∀ {n} (xs : Vec A (1 + n)) →
-           ∃₂ λ (ys : Vec A n) (y : A) → xs ≡ ys ∷ʳ y
-initLast {n = zero}  (x ∷ [])         = ([] , x , refl)
-initLast {n = suc n} (x ∷ xs)         with initLast xs
-initLast {n = suc n} (x ∷ .(ys ∷ʳ y)) | (ys , y , refl) =
-  ((x ∷ ys) , y , refl)
+  filter : ∀[ Vec A ⇒ Bounded A ]
```

Comment:
Perhaps you should write these types out longhand?

Reply:
Do you mean inlining the module parameters?

Reply:
No, I meant write `∀ {n} → Vec A n → Bounded A n`. It's a bit clearer and the `∀[_]` syntax isn't widely used in the library.